### PR TITLE
Generalize socket-based event loop

### PR DESCRIPTION
**Pull request by kpschoedel**
#### Problem

Socket event management is currently tied closely to select(2),
and does not integrate well with other event loops.

See issue #5556 _rework system layer event loop_

#### Change overview

The primary purpose of this change is to provide a clear interface for
monitoring socket events, that can be used both within the CHIP SDK
and by an application using the SDK, and can be implemented either
inside or outside the SDK. Functionality does not change.

* Defines an interface, in `system/SystemSockets.h`.
* Converts most existing uses of socket event monitoring (except mDNS,
  to be done as a followup).
* Converts the existing select(2)-based event management into an
  implementation of the new interface.
* Adds a second implementation of the interface, using libevent,
  to verify that the interface is general enough.

#### Testing

* Converted the applicable unit tests from select(2) to this interface,
  and ran them with both the select and libevent implementations.
  (CI remains select-only for now.)
* Integration tests involving socket-based platforms exercise this code
  and confirm no change to functionality.
* Manual verification of controller/device communication

### CODE REVIEW

File: src/inet/InetLayer.cpp
Comment (gerickson) [inline src/inet/InetLayer.cpp:34]:
```suggestion
 *      For BSD/POSIX Sockets (CHIP_SYSTEM_CONFIG_USE_SOCKETS), event readiness
```

File: src/inet/TCPEndPoint.cpp
Comment (gerickson) [inline src/inet/TCPEndPoint.cpp:284]: It's unclear from the current arrangement of the code whether CHIP_SYSTEM_CONFIG_USE_IO_THREAD == (CHIP_SYSTEM_CONFIG_USE_SOCKETS == 1) && (CHIP_SYSTEM_CONFIG_SOCKETS_USE_SELECT == 0) or if CHIP_SYSTEM_CONFIG_USE_IO_THREAD is completely independent of either of those.

File: src/inet/TCPEndPoint.cpp
Comment (gerickson) [inline src/inet/TCPEndPoint.cpp:2375]: Any reason to not make `IsReadable()` as a convenience inline for `Has(System::SocketEventFlag::kRead)`?

File: src/inet/TCPEndPoint.cpp
Comment (gerickson) [inline src/inet/TCPEndPoint.cpp:2383]: Any reason to not make `IsWriteable()` as a convenience inline for `Has(System::SocketEventFlag::kWrite)`?

File: src/lib/support/SafeInt.h
Comment (gerickson) [inline src/lib/support/SafeInt.h:38]: Probably worth factoring this out as a simple one-liner PR.

File: src/system/SystemConfig.h
Comment (gerickson) [inline src/system/SystemConfig.h:114]: See prior comments about CHIP_SYSTEM_CONFIG_USE_IO_THREAD. There might be an opportunity to detect and flag incorrect configurations.

File: src/system/SystemSocketWatcherSelect.cpp
Comment (gerickson) [inline src/system/SystemSocketWatcherSelect.cpp:53]: See prior comments on convenience inlines.

**Comment by gerickson** [comment]
Will check this out. With CHIP, one of my integration goals would be / is to make the CHIP stack an easy add-on as a CoreFoundation run loop source. I'll be looking at this through that integration lens.

**Comment by gerickson** [comment]
> Will check this out. With CHIP, one of my integration goals would be / is to make the CHIP stack an easy add-on as a CoreFoundation run loop source. I'll be looking at this through that integration lens.

The PR seems reasonable; however, not having ever worked with Darwin libnetwork, I'm probably not the best person to assess that integration goal (but the code doesn't seem harmful toward that end at all and certainly seems to help). As for the CoreFoundation run loop, at minimum, such an integration would seem to require an import/export of sockets, which is likely no different / worse than integrating CoreFoundation run loops with a pure BSD sockets implementation, as opposed to CFNetwork sources.

**Comment by msandstedt** [comment]
I had previously found that for the posix platform, the following was sufficient to allow driving the task function from an existing event loop:

```
diff --git a/src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp b/src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp
index 5d34c339bb..abe2087683 100644
--- a/src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp
+++ b/src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp
@@ -155,8 +155,11 @@ void GenericPlatformManagerImpl_POSIX<ImplClass>::SysProcess()
     nextTimeoutMs = mNextTimeout.tv_sec * 1000 + mNextTimeout.tv_usec / 1000;
     _StartChipTimer(nextTimeoutMs);
 
+    // override; select must not block.
+    struct timeval immediate = { .tv_sec = 0, .tv_usec = 0 };
+
     Impl()->UnlockChipStack();
-    selectRes = select(mMaxFd + 1, &mReadSet, &mWriteSet, &mErrorSet, &mNextTimeout);
+    selectRes = select(mMaxFd + 1, &mReadSet, &mWriteSet, &mErrorSet, &immediate);
     Impl()->LockChipStack();
 
     if (selectRes < 0)
@@ -192,7 +195,7 @@ void GenericPlatformManagerImpl_POSIX<ImplClass>::_RunEventLoop()
     {
         SysUpdate();
         SysProcess();
-    } while (mShouldRunEventLoop.load(std::memory_order_relaxed));
+    } while (0); // mShouldRunEventLoop.load(std::memory_order_relaxed));
 
     Impl()->UnlockChipStack();
 }
```

One of the main sales pitches here is that this set of changes also can allow such a thing.  But if the above is so concise, why are the proposed changes in this PR better?  I'm not asking this as a rhetorical question.  It is implied what's in this PR is better and I want to understand why.

I can see some other problems with select, but I don't think we have a C10K problem to solve.  I'm therefore assuming performance isn't a primary motivator.

**Review by mspang — COMMENTED** [review]
> I had previously found that for the posix platform, the following was sufficient to allow driving the task function from an existing event loop:
> 
> ```
> diff --git a/src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp b/src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp
> index 5d34c339bb..abe2087683 100644
> --- a/src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp
> +++ b/src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp
> @@ -155,8 +155,11 @@ void GenericPlatformManagerImpl_POSIX<ImplClass>::SysProcess()
>      nextTimeoutMs = mNextTimeout.tv_sec * 1000 + mNextTimeout.tv_usec / 1000;
>      _StartChipTimer(nextTimeoutMs);
>  
> +    // override; select must not block.
> +    struct timeval immediate = { .tv_sec = 0, .tv_usec = 0 };
> +
>      Impl()->UnlockChipStack();
> -    selectRes = select(mMaxFd + 1, &mReadSet, &mWriteSet, &mErrorSet, &mNextTimeout);
> +    selectRes = select(mMaxFd + 1, &mReadSet, &mWriteSet, &mErrorSet, &immediate);
>      Impl()->LockChipStack();
>  
>      if (selectRes < 0)
> @@ -192,7 +195,7 @@ void GenericPlatformManagerImpl_POSIX<ImplClass>::_RunEventLoop()
>      {
>          SysUpdate();
>          SysProcess();
> -    } while (mShouldRunEventLoop.load(std::memory_order_relaxed));
> +    } while (0); // mShouldRunEventLoop.load(std::memory_order_relaxed));
>  
>      Impl()->UnlockChipStack();
>  }
> ```

That code does not seem sufficient; I don't even see how you would correctly combine this approach with polling other (non-CHIP) file descriptors on the same thread.

To fix the problem, CHIP must register its file descriptors to be included in application's poll set, so that the thread can block itself pending IO on **either** CHIP owned or application owned file descriptors. Properly there can only be one blocking I/O poll on any given thread. If instead CHIP does a non-blocking poll, and the application does a blocking one, then we won't wake immediately when e.g. UDP packets arrive for CHIP. It may appear to work if there's ongoing activity, but it risks unbounded additional response latency.

This is not about scalability, it's about having the flexibility to use an applications' preferred event loop for the file descriptors managed by CHIP.

> 
> One of the main sales pitches here is that this set of changes also can allow such a thing. But if the above is so concise, why are the proposed changes in this PR better? I'm not asking this as a rhetorical question. It is implied what's in this PR is better and I want to understand why.
> 
> I can see some other problems with select, but I don't think we have a C10K problem to solve. I'm therefore assuming performance isn't a primary motivator.

File: src/system/SystemSocketWatcherLibevent.cpp
Comment (mspang) [inline src/system/SystemSocketWatcherLibevent.cpp:51]: Should this be event_base_loop?

File: src/system/SystemSocketWatcherSelect.h
Comment (mspang) [inline src/system/SystemSocketWatcherSelect.h:69]: Will we able to remove this discrepancy and make both work with SetCallback & remove HandleSelectResult?

File: src/system/SystemSockets.h
Comment (mspang) [inline src/system/SystemSockets.h:51]: We could replace this class with

`using SystemSocketWatcherCallback = void(int socket, SocketEvents events, void * data);` 

?

File: src/inet/TCPEndPoint.cpp
Comment (kpschoedel) [reply src/inet/TCPEndPoint.cpp:284]: Yes, this is a part that needs to be fleshed out. It's independent of the interface for file descriptor watching, but dependent on the implementation. Some may require a thread (as select does), some may require there not be one (e.g. using a CHIP-consuming application's mechanism instead), and some may go either way. Actual `#if` tests should end up much reduced, since I can probably abstract `WakeIOThread()` (previously `WakeSelect()`) a bit more.

File: src/inet/InetLayer.cpp
Comment (kpschoedel) [reply src/inet/InetLayer.cpp:34]: Yes. I will definitely fix and add substantially more documentation.

Comment (kpschoedel) [reply src/inet/InetLayer.cpp:34]: Yes. I will definitely fix and add substantially more documentation.

File: src/inet/TCPEndPoint.cpp
Comment (kpschoedel) [reply src/inet/TCPEndPoint.cpp:2375]: The pattern for most other flag sets in the code is not to have them, so I didn't re-add them when moving `SocketEvents` to use the type-safe `BitFlags` when I had to move it from inet to system. (A bit sad that enum class can't have methods, not to mention a bitwise mixin.)

File: src/lib/support/SafeInt.h
Comment (kpschoedel) [reply src/lib/support/SafeInt.h:38]: Yes. (The reason I changed this actually no longer exists in this PR. _Most_ uses of `CanCastTo` are testing dynamically computer values, but I think there are a small number of tests that could be replaced with `static_assert`.  `CanCastTo` couldn't be `constexpr` in C++11, but can be now in C++14.)

File: src/system/SystemSocketWatcherLibevent.cpp
Comment (kpschoedel) [reply src/system/SystemSocketWatcherLibevent.cpp:51]: Probably. I've gone back and forth on what exactly `HandleEvents()` should mean, so the second opinion is welcome.

File: src/system/SystemSocketWatcherSelect.h
Comment (kpschoedel) [reply src/system/SystemSocketWatcherSelect.h:69]: Yes, I think that should be easy; I just put it off. The remaining #if-select() code would be the `PrepareSelect`/`HandleSelectResult` stuff in //system, which looks a bit harder to refactor, but probably possible — if anybody's still going to want to use select() in the long run.

**Comment by gerickson** [comment]
@mspang and @msandstedt both raise good points here.

Ideally, I can have a single-threaded, simple application with multiple "sources" on the main select loop. CHIP should be one potential integration source in such an application. Perhaps Avahi has some inspiring APIs in this regard in terms of how Avahi can be added as such an integration source. It's been a decade-plus since I've looked at that code; however, I think it amounted to being able to import/export the descriptors Avahi uses for its passive/active sockets and being able to patch those in/out of the main select loop with appropriate callbacks when those particular descriptors have "work" to do.

IIRC, Apple's mDNS/DNS-SD CFNetworkService code works similarly but in a CFRunLoop-centric way (which is just, at its core, a select loop as hand-waved above). There's a main application CFRunLoop and CFNetworkService is patched in as a source of work on that run loop. IIRC, there was/is no need to maintain a separate mDNS/DNS-SD work thread and manage complexities around multiple threads.

If the CHIP stack had the ability, on POSIX-like systems, to export/import descriptors similarly, then it would be possible to have a CFRunLoop adapter, a GLib adapter, etc. for the stack to integrate with each of those run loop types.

**Comment by kpschoedel** [comment]
> If the CHIP stack had the ability, on POSIX-like systems, to export/import descriptors similarly, then it would be possible to have a CFRunLoop adapter, a GLib adapter, etc. for the stack to integrate with each of those run loop types.

That's a good summary of the intent of this. If the `SystemSocketWatcher` is implemented outside CHIP, it's how CHIP exports descriptors; if it's inside, it's how it imports them.

**Comment by gerickson** [comment]
> > If the CHIP stack had the ability, on POSIX-like systems, to export/import descriptors similarly, then it would be possible to have a CFRunLoop adapter, a GLib adapter, etc. for the stack to integrate with each of those run loop types.
> 
> That's a good summary of the intent of this. If the `SystemSocketWatcher` is implemented outside CHIP, it's how CHIP exports descriptors; if it's inside, it's how it imports them.

I am working on a personal project that I am looking to open source any day now in which I'd like to place a CHIP front-end for the proxy aspect of the project. The timeline may be a bit too latent (probably 1-3 months out) for CHIP 1.0; however, I'd be happy to work on what a CFRunLoop adapter would need to look like to help with the effort.

**Comment by mspang** [comment]
LGTM

File: src/system/SystemSockets.h
Comment (kpschoedel) [reply src/system/SystemSockets.h:51]: Actually, after further investigation, I do want to have a base class containing the file descriptor and active-event flags — essentially, changing the consumers' members from `int mSocket; SocketEvents mPendingIO; SystemSocketWatcher mWatcher;` as here, to `WatchableSocket mSocket`.

**Review by woody-apple — DISMISSED** [review]
Per the updated template, can you update the PR here?

```
#### Problem
What is being fixed?  Examples:
* Fix crash on startup
* Fixes #12345 12345 Frobnozzle is leaky (exactly like that, so GitHub will auto-close the issue).

#### Change overview
What's in this PR

#### Testing
How was this tested? (at least one bullet point required)
    • If unit tests were added, how do they cover this issue?
    • If unit tests existed, how were they fixed/modified to prevent this in future?
    • If integration tests were added, how do they verify this change?
    • If manually tested, what platforms controller and device platforms were manually tested, and how?
    • If no testing is required, why not?
```

**Comment by gerickson** [comment]
> > > If the CHIP stack had the ability, on POSIX-like systems, to export/import descriptors similarly, then it would be possible to have a CFRunLoop adapter, a GLib adapter, etc. for the stack to integrate with each of those run loop types.
> > 
> > 
> > That's a good summary of the intent of this. If the `SystemSocketWatcher` is implemented outside CHIP, it's how CHIP exports descriptors; if it's inside, it's how it imports them.
> 
> I am working on a personal project that I am looking to open source any day now in which I'd like to place a CHIP front-end for the proxy aspect of the project. The timeline may be a bit too latent (probably 1-3 months out) for CHIP 1.0; however, I'd be happy to work on what a CFRunLoop adapter would need to look like to help with the effort. I'll post a link to the project when it's been open sourced.

As promised: https://github.com/gerickson/openhlx/. I’ll be attempting to plumb CHIP/Matter into the nascent hlxproxyd.

File: src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp
Comment (mspang) [inline src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp:148]: Are we going to be able to replace this code with the Android implementation https://developer.android.com/reference/android/os/Looper ?

File: src/system/WatchableSocketSelect.cpp
Comment (mspang) [inline src/system/WatchableSocketSelect.cpp:87]: It is a misconception that exceptfds is related to the occurrence of errors.

**Review by mspang — APPROVED** [review]

File: src/system/SystemTimer.cpp
Comment (kpschoedel) [reply src/system/SystemTimer.cpp:193]: I'm not sure yet (and haven't read the Darwin dispatch system yet), but this is one reason I've deferred anything but minimal timer changes. Obviously for select(2) they're tightly integrated but an implementation shouldn't have to do that. Likely the substitutable implementation part (what I've called `WatchableEventManager` here) will get timer methods, but details TBD.

File: src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp
Comment (kpschoedel) [reply src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp:148]: I don't know any reason we shouldn't be able to, since the consuming-side interface (i.e. SystemSockets.h) is general. As long as there's some way to implement calling back on events of interest, it should be doable.

File: src/system/WatchableSocketSelect.cpp
Comment (kpschoedel) [reply src/system/WatchableSocketSelect.cpp:87]: True, that's a holdover from inherited code. I think `SocketEvents` can get a fourth flag for clarity at no additional cost.

Comment (mspang) [reply src/system/WatchableSocketSelect.cpp]: Ack

File: src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp
Comment (mspang) [reply src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp:148]: Ok sounds good, I don't expect to solve it in this PR. I do think Android will want to move onto their Java based event loop.

Comment (mspang) [reply src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp:148]: @austinh0

File: src/system/SystemTimer.cpp
Comment (mspang) [reply src/system/SystemTimer.cpp:193]: @vivien-apple

**Comment by mspang** [comment]
@msandstedt @bzbarsky-apple @woody-apple

File: src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp
Comment (austinh0) [reply src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp:180]: Filed #7626.

**Comment by mspang** [comment]
@bzbarsky-apple @msandstedt

**Review by msandstedt — APPROVED** [review]

**Comment by kpschoedel** [comment]
> Per the updated template, can you update the PR here?

Description is updated

**Comment by andy31415** [comment]
@woody-apple - dismissed your change request on the assumption it was for the summary formatting. Please re-review when you have a chance.

**Comment by woody-apple** [comment]
@bzbarsky-apple can you review?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/system/SystemSockets.h
Comment (kpschoedel) [reply src/system/SystemSockets.h:46]: These are just copied from the current `class SocketEvents`. I personally agree with removing the prefix, but a large majority of enum class currently use it (arguably _also_ my fault, I guess, since I did the conversion from plain enums for many.)

File: src/system/SystemSockets.h
Comment (kpschoedel) [reply src/system/SystemSockets.h:251]: Added to my followup to-do list.

File: src/system/SystemSockets.h
Comment (kpschoedel) [reply src/system/SystemSockets.h:271]: Also just copied from current code (currently in `system/SystemWakeEvent.h`). Will rename in followup.

Comment (kpschoedel) [reply src/system/SystemSockets.h:271]: Copied directly from existing <system/SystemWakeEvent.h>; will add documentation in followup.

File: src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp
Comment (kpschoedel) [reply src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp:134]: Correct. I guess I considered fixing the comments to part of the TODO.

File: src/inet/RawEndPoint.cpp
Comment (kpschoedel) [reply src/inet/RawEndPoint.cpp:428]: Whether the equivalent is needed or not (and whether a thread exists or not) depends on the event-watching implementation. It's potentially needed once there is an actual request for notification, so in this function that's by `mSocket.RequestCallbackOnPendingRead()` below; in the select(2)-based implementation that ends up waking the thread if we weren't already waiting on this socket.

File: src/inet/RawEndPoint.cpp
Comment (kpschoedel) [reply src/inet/RawEndPoint.cpp:481]: Similarly, `mSocket.Close()` ends up waking the thread if the implementation requires it, via `WatchableSocket::OnClose()`. The select version does it there; the libevent version doesn't need to.

File: src/system/SystemSockets.h
Comment (kpschoedel) [reply src/system/SystemSockets.h:271]: Will rename now just to kick the CI.

**Pull request merged**
